Add SQLite Wasm article (#4620)
GoogleChrome/developer.chrome.com: site/en/blog/sqlite-wasm-in-the-browser-backed-by-the-origin-private-file-system/index.md

is our fulfillment of this promise. {% endAside %}

## About SQLite

[SQLite](https://sqlite.org/) is a popular, [open-source](https://sqlite.org/src/tree), lightweight,
embedded relational database management system. Many developers use it to store data in a
structured, easy-to-use manner. Because of its small size and low memory requirements, SQLite is
often leveraged as a database engine in mobile devices, desktop applications, and web browsers.

One of the key features of SQLite is that it is a serverless database, which means that it does not
require a separate server process to operate. Instead, the database is stored in a single file on
the user's device, making it easy to integrate into applications.

{% Img src="image/8WbTDNrhLsU0El80frMBGE4eMCD3/eDfwxFJirAZ6hqdt6Z8Y.png", alt="SQLite logo.", width="220", height="101" %}

## SQLite based on Web Assembly

There are a number of unofficial SQLite versions based on Web Assembly (Wasm), allowing it to be
used in web browsers, for example, [sql.js](https://github.com/sql-js/sql.js). The
[sqlite3 WASM/JS subproject](https://sqlite.org/wasm/doc/tip/about.md) is the first effort that is

> "officially" associated with the [SQLite project](https://sqlite.org/) makingWasm builds of the

Reviewer: ```suggestion
"officially" associated with the [SQLite project](https://sqlite.org/) making Wasm builds of the
```

Reviewer: also why the scare quotes round officially?

Author: Quotes be gone.

- A Promise-based variant of the Worker API which entirely hides the cross-thread communication
  aspects from the user.
- Support for persistent client-side storage using available JavaScript APIs, including the Origin
  Private File System (OPFS).

## Using SQLite Wasm with the Origin Private File System persistence backend

### Downloading the library

Download the official Wasm build from the
[SQLite homepage](https://sqlite.org//download.html#:~:text=WebAssembly%20%26%20JavaScript). Be sure
to download the right build. The description needs to mention `sqlite3.wasm`. After the download has
finished, extract the archive.

### The Origin Private File System

The Origin Private File System (OPFS, part of the
[File System Access API](/articles/file-system-access/)) is augmented with a special surface that
brings very performant access to data. This new surface differs from existing ones by offering
in-place and exclusive write access to a file's content. This change, along with the ability to
consistently read unflushed modifications and the availability of a synchronous variant on dedicated
workers, significantly improves performance and unblocks new use cases.

> As you can imagine, the last point of the project's goals, _Support for persistent client-side
> storage using available JavaScript APIs_ specifically, comes with strict performance requirements

Reviewer: ```suggestion
As you can imagine, the last point of the project's goals, Support for persistent client-side
storage using available JavaScript APIs, comes with strict performance requirements
```

regarding persisting data to the database file. This is where the Origin Private File System, and,
more specifically, the
[`createSyncAccessHandle()`](https://developer.mozilla.org/docs/Web/API/FileSystemFileHandle/createSyncAccessHandle)
method of [`FileSystemFileHandle`](https://developer.mozilla.org/docs/Web/API/FileSystemFileHandle)
objects comes into play. This method returns a Promise which resolves to a
[`FileSystemSyncAccessHandle`](https://developer.mozilla.org/docs/Web/API/FileSystemSyncAccessHandle)
object that can be used to synchronously read from and write to a file. The synchronous nature of
this method brings performance advantages, but therefore it is only usable inside dedicated
[Web Workers](https://developer.mozilla.org/docs/Web/API/Web_Workers_API) for files within the
Origin Private File System so the main thread can't be blocked.

### Setting the required headers

> {% Aside 'warning' %} Setting these headers is a _required_ step for the Origin Private File System

Reviewer: ```suggestion
{% Aside 'warning' %} Setting these headers is a required step for the Origin Private File System
```
